# Milestone 2 DSCI 525

Group 18

## GitHub Repository

You can access the GitHub Repository with this [link](https://github.com/UBC-MDS/525_group18).

## Milestone 2 checklist  

- [x] Setup your EC2 instance with JupyterHub.  
- [x] Install all necessary things needed in your UNIX server (amazon ec2 instance).
- [x] Setup your S3 bucket.
- [x] Move the data that you wrangled in your last milestone to s3.
- [x] Get the data from S3 in your notebook and make data ready for machine learning.

**Keep in mind:**

- _All services you use are in region Canada ca-central-1._

- _Use only default VPC and subnet, also if not specified explicitly in instruction, leave those options as default when setting up an instance and s3._
    
- _No IP addresses are visible when you provide the screenshot._

- _We want one single notebook for grading, and it's up to your discretion on how you do it. ***So only person in your group spins ```t2.xlarge```.***_

- _Say something went wrong and you want to spin up another EC2 instance, then make sure you terminate the previous one._

- _We will be choosing the storage to be ```Delete on Termination```, which means upon termination, stored data in your instance will be lost. Make sure you save any data to S3 and download the notebooks to your laptop so that next time you have your jupyterHub in a different instance, you can upload your notebook there._

_***Outside of Milestone:*** If you are working as an individual to just practice setting up EC2 instances, make sure you select ```t2.micro``` instance and stick to other ```settings/instructions``` given in this milestone. I strongly recommend you spinning up your own instance and experimenting with the s3 bucket in doing something to get comfortable with AWS. But we won't be looking at it for a grading purpose._

### 1. Setup your EC2 instance

rubric={correctness:20}

Follow the instructions shown during the lecture to set up your EC2 instance. You can use [this](https://tljh.jupyter.org/en/latest/install/amazon.html ) as your reference, but please make sure you follow the below instructions.

1.1) Choose AMI "Ubuntu Server 18.04 LTS (HVM), SSD Volume Type 64-bit (x86)".

1.2) Choose an Instance Type t2.xlarge.

1.2) Make sure you go with the default VPC & subnet.

1.4) Get the configuration code from step 7 in the above link and replace "admin-user-name"(remove < > as well) with your AWS IAM username.
    
1.5) Storage use Root with size 30 GB.
    
1.6) Add tag, enter "Owner" under the Key field. In the Value field in the Name row, give your IAM username.
    
1.7) Select an existing security group Name "DSCI525."
    
1.8) Review page looks like [this](https://github.ubc.ca/MDS-2020-21/DSCI_525_web-cloud-comp_students/blob/master/images/1_out.png) before you launch the instance.
    
1.9) In the pop-up, "Select an existing key pair or a new key pair." If you are setting up your cluster for the first time, click on create a new key pair and name it as your "IAM user account". Download the private key and keep it secure. Next time when you set up the EC2 instance, make sure you select "Choose an existing keypair" and choose the one you already made for you.
    
1.10) Search for your "IAM user account" under instances to see if it's running. Give it 15 - 20 min as even if it shows running, it will take more time to set up JupyterHub. So please wait...!
    
1.11) Check out the "Connect" button to determine how you can connect to the instance. Now you are given the DOOR access to the server, which we mentioned in our first class. :)

#### Screenshot for Step 1

https://raw.githubusercontent.com/UBC-MDS/525_group18/main/imgs/Milestone2/Screen%20Shot%202021-04-16%20at%205.28.46%20PM.png

### 2. Setup your JupyterHub

rubric={correctness:20}

2.1) Under description, check for "IPv4 Public IP" and paste the IP address in your browser for your JupyterHub.

2.2) Enter your "IAM user account" and use a strong password & note it down somewhere, as what you enter here will be the admin password.
    
2.3) In your JupyterHub, go to "Control Panel" --> "admin." Here add other members of your group use their "IAM user account" and make them admins.
    
2.4) Check if other members can log in to the JupyterHub from their machines by giving them the URL to connect. Step 2.2 is applicable here for other members.

#### Screenshot for Step 2

https://raw.githubusercontent.com/UBC-MDS/525_group18/main/imgs/Milestone2/Screen%20Shot%202021-04-16%20at%205.36.06%20PM.png

### 3. Setup the server 

rubric={correctness:20}

3.1) Login in to the server (instance). The person who spins up the EC2 instance will only have access to the server as he only got the private key. If someone else wants to log in to that instance, you need to get hold of that private key (  Refer 1.10 ). Need to know more ? [Click here](https://github.ubc.ca/MDS-2020-21/DSCI_521_platforms-dsci_instructors/blob/master/lectures/7-ssh-filenames-project-organization.ipynb)

3.2) Setup a common data folder to download data, and this folder should be accessible by all users in the JupyterHub. Following commands make a folder and make it accessible to everyone. Want to learn more about basic UNIX commands? [Click here](https://maker.pro/linux/tutorial/basic-linux-commands-for-beginners).

```
sudo mkdir -p /srv/data/my_shared_data_folder
sudo chmod 777 /srv/data/my_shared_data_folder/
```
    
3.3)(***OPTIONAL, no bonus points***) If you want a sharing notebook environment, then check out [this]( https://github.com/gutow/tljh_grp_utils). if you plan to do this, make sure you install the "members" package in your server run ```sudo apt-get install members```."

3.4) Install AWS CLI. More details [here](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2-linux.html).

_NOTE:We are installing this in our EC2 instance, but we can install this anywhere to interact with s3. Say you can install it in your local machine and move data to s3._

    curl "https://awscli.amazonaws.com/awscli-exe-linux-x86_64.zip" -o "awscliv2.zip"
    sudo apt install unzip
    unzip awscliv2.zip
    sudo ./aws/install
    
3.5) Setup your access key and secret. Do it from your [AWS console](https://console.aws.amazon.com/iam/home?#/security_credentials). Make sure you keep your "Access key ID" & secret key somewhere safe.

3.6) Use these credentials to configure AWS CLI (aws configure). More details [here](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-quickstart.html). "Default region" and "output format" you can leave empty.

3.7) AWS cli can be used to interact with a lot of services. Check this [out](https://docs.aws.amazon.com/cli/latest/reference/). To get a feel, we will use CLI to interact with s3 and wait for step 6.

#### Screenshot for step 3

https://raw.githubusercontent.com/UBC-MDS/525_group18/main/imgs/Milestone2/Screen%20Shot%202021-04-16%20at%205.43.16%20PM.png

### 4. Get the data what we wrangled in our first milestone. 

#### Library Imports

In [3]:


import re
import os
import glob
import zipfile
import requests
from urllib.request import urlretrieve
import json
import pandas as pd
from memory_profiler import memory_usage


In [4]:


%load_ext memory_profiler



The memory_profiler extension is already loaded. To reload it, use:
  %reload_ext memory_profiler


#### Downloading the Data

In [5]:


# Necessary metadata
article_id = 14226968  # this is the unique identifier of the article on figshare
url = f"https://api.figshare.com/v2/articles/{article_id}"
headers = {"Content-Type": "application/json"}
output_directory = "/srv/data/my_shared_data_folder/"



In [6]:
response = requests.request("GET", url, headers=headers)
data = json.loads(response.text)  # this contains all the articles data, feel free to check it out
files = data["files"]             # this is just the data about the files, which is what we want
files

[{'is_link_only': False,
  'name': 'allyears.csv.zip',
  'supplied_md5': '9e046ac05ecd2c32a256a47dd1098b81',
  'computed_md5': '9e046ac05ecd2c32a256a47dd1098b81',
  'id': 26844650,
  'download_url': 'https://ndownloader.figshare.com/files/26844650',
  'size': 2405908113},
 {'is_link_only': False,
  'name': 'individual_years.zip',
  'supplied_md5': '921da748974b07b2a70bbfcc04535a77',
  'computed_md5': '921da748974b07b2a70bbfcc04535a77',
  'id': 26863682,
  'download_url': 'https://ndownloader.figshare.com/files/26863682',
  'size': 1896206676},
 {'is_link_only': False,
  'name': 'combined_model_data.csv.zip',
  'supplied_md5': '7638434c44a7d29cbb29fe200b4fd65d',
  'computed_md5': '7638434c44a7d29cbb29fe200b4fd65d',
  'id': 27515426,
  'download_url': 'https://ndownloader.figshare.com/files/27515426',
  'size': 821308997},
 {'is_link_only': False,
  'name': 'combined_model_data_parti.parquet.zip',
  'supplied_md5': '02f4e3df8d16580a02291de225072689',
  'computed_md5': '02f4e3df8d16580a02

In [6]:

files_to_dl = ["combined_model_data_parti.parquet.zip"]  ## Please download the partitioned 
for file in files:
    if file["name"] in files_to_dl:
        os.makedirs(output_directory, exist_ok=True)
        urlretrieve(file["download_url"], output_directory + file["name"])



In [7]:


with zipfile.ZipFile(os.path.join(output_directory, "combined_model_data_parti.parquet.zip"), 'r') as f:
    f.extractall(output_directory)



### 5. Setup your S3 bucket and move data

rubric={correctness:20}

5.1) Get comfortable with S3 UI. Go from the AWS console.

5.2) Create a bucket there. The name should be mds-s3-xxx. Replace xxx with your "IAM user account".

5.3) All other options leave as it is. (Make sure AWS region is Canada).

5.4) Create your first folder called "output".

5.5) Move the "observed_daily_rainfall_SYD.csv" file from the Milestone1 data folder to your s3 bucket from your local computer. (it's a tiny file, so maybe you can easily use UI to upload).

5.6) Moving the parquet file we downloaded in step 4 to S3 using the cli what we installed in step 3.7.
Refer [this](https://docs.aws.amazon.com/cli/latest/userguide/cli-services-s3-commands.html#using-s3-commands-managing-objects-move) document and figure out yourself!
    
Hint: We are interested in the ```cp``` command. ```local``` is the directory path on our server.

#### Screenshot for step 5

https://raw.githubusercontent.com/UBC-MDS/525_group18/main/imgs/Milestone2/Screen%20Shot%202021-04-16%20at%205.55.56%20PM.png

### 6. Wrangle the data in preparation for machine learning

In [7]:


%%time
%%memit

# Passing your credentials 
## IMPORTANT: make sure you dont include you secret and key when submitting the notebook
aws_credentials ={"key": "use your own","secret": "use yur own"} ## dont include you secret and key when submitting the notebook
df = pd.read_parquet("s3://mds-s3-student65/combined_model_data_parti.parquet", storage_options=aws_credentials)


peak memory: 7307.52 MiB, increment: 7211.48 MiB
CPU times: user 10.2 s, sys: 5.73 s, total: 16 s
Wall time: 30.5 s


In [45]:
df.head()

time  lat_min  lat_max  lon_min  lon_max  rain (mm/day)  \
0 1889-01-01 12:00:00   -36.25    -35.0  140.625    142.5   3.293256e-13   
1 1889-01-02 12:00:00   -36.25    -35.0  140.625    142.5   0.000000e+00   
2 1889-01-03 12:00:00   -36.25    -35.0  140.625    142.5   0.000000e+00   
3 1889-01-04 12:00:00   -36.25    -35.0  140.625    142.5   0.000000e+00   
4 1889-01-05 12:00:00   -36.25    -35.0  140.625    142.5   1.047658e-02   

        model  
0  ACCESS-CM2  
1  ACCESS-CM2  
2  ACCESS-CM2  
3  ACCESS-CM2  
4  ACCESS-CM2

In [46]:
df

time  lat_min  lat_max  lon_min  lon_max  \
0        1889-01-01 12:00:00   -36.25    -35.0  140.625    142.5   
1        1889-01-02 12:00:00   -36.25    -35.0  140.625    142.5   
2        1889-01-03 12:00:00   -36.25    -35.0  140.625    142.5   
3        1889-01-04 12:00:00   -36.25    -35.0  140.625    142.5   
4        1889-01-05 12:00:00   -36.25    -35.0  140.625    142.5   
...                      ...      ...      ...      ...      ...   
62513858 2014-12-27 00:00:00      NaN      NaN      NaN      NaN   
62513859 2014-12-28 00:00:00      NaN      NaN      NaN      NaN   
62513860 2014-12-29 00:00:00      NaN      NaN      NaN      NaN   
62513861 2014-12-30 00:00:00      NaN      NaN      NaN      NaN   
62513862 2014-12-31 00:00:00      NaN      NaN      NaN      NaN   

          rain (mm/day)       model  
0          3.293256e-13  ACCESS-CM2  
1          0.000000e+00  ACCESS-CM2  
2          0.000000e+00  ACCESS-CM2  
3          0.000000e+00  ACCESS-CM2  
4          1.047658e-02  ACCESS-CM2  
...                 ...         ...  
62513858   3.747200e-02    observed  
62513859   1.580613e-01    observed  
62513860   2.571914e-02    observed  
62513861   7.293899e-01    observed  
62513862   8.076387e-03    observed  

[62513863 rows x 7 columns]

In [47]:
df.shape

(62513863, 7)

Now let us proceed with wrangling the rainfall dataframe to only include records for Sydney, Australia.

syd_lat = -33.86

syd_lon = 151.21

In [48]:
syd_lat = -33.86

syd_lon = 151.21

In [49]:
sydney = df.query("lat_min < @syd_lat < lat_max & lon_min < @syd_lon < lon_max")
sydney

time    lat_min    lat_max  lon_min  lon_max  \
552240   1889-01-01 12:00:00 -35.000000 -33.750000  150.000  151.875   
552241   1889-01-02 12:00:00 -35.000000 -33.750000  150.000  151.875   
552242   1889-01-03 12:00:00 -35.000000 -33.750000  150.000  151.875   
552243   1889-01-04 12:00:00 -35.000000 -33.750000  150.000  151.875   
552244   1889-01-05 12:00:00 -35.000000 -33.750000  150.000  151.875   
...                      ...        ...        ...      ...      ...   
60352298 2014-12-27 12:00:00 -33.926702 -32.984293  150.625  151.875   
60352299 2014-12-28 12:00:00 -33.926702 -32.984293  150.625  151.875   
60352300 2014-12-29 12:00:00 -33.926702 -32.984293  150.625  151.875   
60352301 2014-12-30 12:00:00 -33.926702 -32.984293  150.625  151.875   
60352302 2014-12-31 12:00:00 -33.926702 -32.984293  150.625  151.875   

          rain (mm/day)       model  
552240         0.040427  ACCESS-CM2  
552241         0.073777  ACCESS-CM2  
552242         0.232656  ACCESS-CM2  
552243         0.911319  ACCESS-CM2  
552244         0.698013  ACCESS-CM2  
...                 ...         ...  
60352298      17.444923     TaiESM1  
60352299       1.569647     TaiESM1  
60352300       1.444630     TaiESM1  
60352301       0.716019     TaiESM1  
60352302       3.240185     TaiESM1  

[1150049 rows x 7 columns]

Futhermore, now let us proceed with dropping the columns that are not needed in the analysis, such as latitude and longtitude values for the city Sydney.

In [50]:
sydney_df = sydney.drop(columns = ["lat_min", "lat_max", "lon_min", "lon_max"])
sydney_df

time  rain (mm/day)       model
552240   1889-01-01 12:00:00       0.040427  ACCESS-CM2
552241   1889-01-02 12:00:00       0.073777  ACCESS-CM2
552242   1889-01-03 12:00:00       0.232656  ACCESS-CM2
552243   1889-01-04 12:00:00       0.911319  ACCESS-CM2
552244   1889-01-05 12:00:00       0.698013  ACCESS-CM2
...                      ...            ...         ...
60352298 2014-12-27 12:00:00      17.444923     TaiESM1
60352299 2014-12-28 12:00:00       1.569647     TaiESM1
60352300 2014-12-29 12:00:00       1.444630     TaiESM1
60352301 2014-12-30 12:00:00       0.716019     TaiESM1
60352302 2014-12-31 12:00:00       3.240185     TaiESM1

[1150049 rows x 3 columns]

Now, let's proceed with setting the time as the index of the dataframe.

In [51]:
sydney_index = sydney_df.set_index("time")
sydney_index

rain (mm/day)       model
time                                          
1889-01-01 12:00:00       0.040427  ACCESS-CM2
1889-01-02 12:00:00       0.073777  ACCESS-CM2
1889-01-03 12:00:00       0.232656  ACCESS-CM2
1889-01-04 12:00:00       0.911319  ACCESS-CM2
1889-01-05 12:00:00       0.698013  ACCESS-CM2
...                            ...         ...
2014-12-27 12:00:00      17.444923     TaiESM1
2014-12-28 12:00:00       1.569647     TaiESM1
2014-12-29 12:00:00       1.444630     TaiESM1
2014-12-30 12:00:00       0.716019     TaiESM1
2014-12-31 12:00:00       3.240185     TaiESM1

[1150049 rows x 2 columns]

It is it time for us to tranform the data so that is appropriate for training a machine learning model. Hence we are going to pivot the column `model` and assign the values of `rain (mm/day)` for each model in study.

In [52]:
sydney_pivot = sydney_index.pivot(columns='model', values = "rain (mm/day)")
sydney_pivot

model                ACCESS-CM2  ACCESS-ESM1-5  AWI-ESM-1-1-LR  BCC-CSM2-MR  \
time                                                                          
1889-01-01 00:00:00         NaN            NaN             NaN          NaN   
1889-01-01 12:00:00    0.040427       1.814552    3.557934e+01     4.268112   
1889-01-02 00:00:00         NaN            NaN             NaN          NaN   
1889-01-02 12:00:00    0.073777       0.303965    4.596520e+00     1.190141   
1889-01-03 00:00:00         NaN            NaN             NaN          NaN   
...                         ...            ...             ...          ...   
2014-12-29 12:00:00    0.005964       3.041667    2.898325e+00     0.093595   
2014-12-30 00:00:00         NaN            NaN             NaN          NaN   
2014-12-30 12:00:00    0.000028       1.131412    2.516381e-01     0.171503   
2014-12-31 00:00:00         NaN            NaN             NaN          NaN   
2014-12-31 12:00:00    0.532747       2.370896    1.047835e-13     4.437736   

model                    BCC-ESM1  CMCC-CM2-HR4  CMCC-CM2-SR5  CMCC-ESM2  \
time                                                                       
1889-01-01 00:00:00           NaN           NaN           NaN        NaN   
1889-01-01 12:00:00  1.107466e-03     11.410537  3.322009e-08   2.668800   
1889-01-02 00:00:00           NaN           NaN           NaN        NaN   
1889-01-02 12:00:00  1.015323e-04      4.014984  1.312700e+00   0.946211   
1889-01-03 00:00:00           NaN           NaN           NaN        NaN   
...                           ...           ...           ...        ...   
2014-12-29 12:00:00  2.000051e-08      0.253221  1.306046e+00   0.000002   
2014-12-30 00:00:00           NaN           NaN           NaN        NaN   
2014-12-30 12:00:00  7.191735e-05      0.081693  1.722262e-01   0.788577   
2014-12-31 00:00:00           NaN           NaN           NaN        NaN   
2014-12-31 12:00:00  2.863683e-01      6.343592  6.368303e-01   0.442130   

model                  CanESM5  EC-Earth3-Veg-LR  ...    MIROC6  \
time                                              ...             
1889-01-01 00:00:00        NaN               NaN  ...       NaN   
1889-01-01 12:00:00   1.321215          1.515293  ...  0.648031   
1889-01-02 00:00:00        NaN               NaN  ...       NaN   
1889-01-02 12:00:00   2.788724          4.771375  ...  1.099883   
1889-01-03 00:00:00        NaN               NaN  ...       NaN   
...                        ...               ...  ...       ...   
2014-12-29 12:00:00  37.169669          1.565885  ...  0.709866   
2014-12-30 00:00:00        NaN               NaN  ...       NaN   
2014-12-30 12:00:00   7.361246          0.025749  ...  0.327474   
2014-12-31 00:00:00        NaN               NaN  ...       NaN   
2014-12-31 12:00:00   0.306608          0.106525  ...  0.042272   

model                MPI-ESM-1-2-HAM  MPI-ESM1-2-HR  MPI-ESM1-2-LR  \
time                                                                 
1889-01-01 00:00:00              NaN            NaN            NaN   
1889-01-01 12:00:00     4.244226e-13   1.390174e-13   6.537884e-05   
1889-01-02 00:00:00              NaN            NaN            NaN   
1889-01-02 12:00:00     4.409552e+00   1.222283e-01   1.049131e-13   
1889-01-03 00:00:00              NaN            NaN            NaN   
...                              ...            ...            ...   
2014-12-29 12:00:00     2.016156e+01   1.506439e+00   1.049481e-01   
2014-12-30 00:00:00              NaN            NaN            NaN   
2014-12-30 12:00:00     9.420543e+00   6.242895e+00   1.245115e-01   
2014-12-31 00:00:00              NaN            NaN            NaN   
2014-12-31 12:00:00     1.031899e+01   4.765813e+01   3.274323e-01   

model                  MRI-ESM2-0      NESM3    NorESM2-LM  NorESM2-MM  \
time                                                                     
1889-01-01 00:00:00           NaN        NaN        

After pivoting the data we see that are missing values, hence let us resample the data and aggragate it with the `mean` function.

In [53]:
df_sydney = sydney_pivot.resample("1D").mean()
df_sydney

model       ACCESS-CM2  ACCESS-ESM1-5  AWI-ESM-1-1-LR   BCC-CSM2-MR  \
time                                                                  
1889-01-01    0.040427       1.814552    3.557934e+01  4.268112e+00   
1889-01-02    0.073777       0.303965    4.596520e+00  1.190141e+00   
1889-01-03    0.232656       0.019976    5.927467e+00  1.003845e-09   
1889-01-04    0.911319      13.623777    8.029624e+00  8.225225e-02   
1889-01-05    0.698013       0.021048    2.132686e+00  2.496841e+00   
...                ...            ...             ...           ...   
2014-12-27    0.033748       0.123476    1.451179e+00  3.852845e+01   
2014-12-28    0.094198       2.645496    4.249335e+01  5.833801e-01   
2014-12-29    0.005964       3.041667    2.898325e+00  9.359547e-02   
2014-12-30    0.000028       1.131412    2.516381e-01  1.715028e-01   
2014-12-31    0.532747       2.370896    1.047835e-13  4.437736e+00   

model           BCC-ESM1  CMCC-CM2-HR4  CMCC-CM2-SR5  CMCC-ESM2    CanESM5  \
time                                                                         
1889-01-01  1.107466e-03  1.141054e+01  3.322009e-08   2.668800   1.321215   
1889-01-02  1.015323e-04  4.014984e+00  1.312700e+00   0.946211   2.788724   
1889-01-03  1.760345e-05  9.660565e+00  9.103720e+00   0.431999   0.003672   
1889-01-04  1.808932e-01  3.951528e+00  1.317160e+01   0.368693   0.013578   
1889-01-05  4.708019e-09  2.766362e+00  1.822940e+01   0.339267   0.002468   
...                  ...           ...           ...        ...        ...   
2014-12-27  2.061717e-03  8.179260e-09  1.171263e-02   0.090786  59.895053   
2014-12-28  5.939502e-09  8.146937e-01  4.938899e-01   0.000000   0.512632   
2014-12-29  2.000051e-08  2.532205e-01  1.306046e+00   0.000002  37.169669   
2014-12-30  7.191735e-05  8.169252e-02  1.722262e-01   0.788577   7.361246   
2014-12-31  2.863683e-01  6.343592e+00  6.368303e-01   0.442130   0.306608   

model       EC-Earth3-Veg-LR  ...        MIROC6  MPI-ESM-1-2-HAM  \
time                          ...                                  
1889-01-01          1.515293  ...  6.480307e-01     4.244226e-13   
1889-01-02          4.771375  ...  1.099883e+00     4.409552e+00   
1889-01-03          4.233980  ...  3.931751e+00     2.269300e-01   
1889-01-04         15.252495  ...  6.273648e-01     2.344586e-02   
1889-01-05         11.920356  ...  1.977652e-02     4.270161e-13   
...                      ...  ...           ...              ...   
2014-12-27          5.071783  ...  2.916075e-11     4.726998e-13   
2014-12-28          1.578188  ...  9.565743e-01     4.609420e-13   
2014-12-29          1.565885  ...  7.098663e-01     2.016156e+01   
2014-12-30          0.025749  ...  3.274740e-01     9.420543e+00   
2014-12-31          0.106525  ...  4.227210e-02     1.031899e+01   

model       MPI-ESM1-2-HR  MPI-ESM1-2-LR    MRI-ESM2-0         NESM3  \
time                                                                   
1889-01-01   1.390174e-13   6.537884e-05  3.445495e-06  1.576096e+01   
1889-01-02   1.222283e-01   1.049131e-13  4.791993e-09  3.675510e-01   
1889-01-03   3.762301e-01   9.758706e-14  6.912302e-01  1.562869e-01   
1889-01-04   4.214019e-01   7.060915e-03  3.835721e-02  2.472226e-07   
1889-01-05   1.879692e-01   4.504985e+00  3.506923e-07  1.949792e-13   
...                   ...            ...           ...           ...   
2014-12-27   1.326889e-01   1.827857e+00  6.912632e-03  2.171327e-03   
2014-12-28   1.644482e+00   7.242920e-01  2.836752e-03  1.344768e+01   
2014-12-29   1.506439e+00   1.049481e-01  8.137182e+00  2.547820e+01   
2014-12-30   6.242895e+00   1.245115e-01  9.305263e-03  4.192948e+00   
2014-12-31   4.765813e+01   3.274323e-01  6.854985e-11  2.067847e+00   

model         NorESM2-LM  NorESM2-MM  SAM0-UNICON    TaiESM1  
time                                                          
1889-01-01  4.759651e-05    2.451075     0.221324   2.257933  
1889-01-02  4.350863e-01    0.477231     3.757179 

Now we need to add an extra column for the target, which is saved in the s3 bucket. The file is called `observed_daily_rainfall_SYD.csv`, let us load the for this.

In [54]:
observed_df = pd.read_csv("s3://mds-s3-student65/observed_daily_rainfall_SYD.csv", storage_options=aws_credentials)

In [55]:
observed_df

time  rain (mm/day)
0      1889-01-01       0.006612
1      1889-01-02       0.090422
2      1889-01-03       1.401452
3      1889-01-04      14.869798
4      1889-01-05       0.467628
...           ...            ...
46015  2014-12-27       0.037472
46016  2014-12-28       0.158061
46017  2014-12-29       0.025719
46018  2014-12-30       0.729390
46019  2014-12-31       0.008076

[46020 rows x 2 columns]

In [56]:
observed_df.shape

(46020, 2)

Let us add this the observed rainfall data as a column in our wrangled Sydney dat set of rainfall values for different models.

In [67]:
df_sydney["observed"] = observed_df[["rain (mm/day)"]].values

In [68]:
df_sydney

model       ACCESS-CM2  ACCESS-ESM1-5  AWI-ESM-1-1-LR   BCC-CSM2-MR  \
time                                                                  
1889-01-01    0.040427       1.814552    3.557934e+01  4.268112e+00   
1889-01-02    0.073777       0.303965    4.596520e+00  1.190141e+00   
1889-01-03    0.232656       0.019976    5.927467e+00  1.003845e-09   
1889-01-04    0.911319      13.623777    8.029624e+00  8.225225e-02   
1889-01-05    0.698013       0.021048    2.132686e+00  2.496841e+00   
...                ...            ...             ...           ...   
2014-12-27    0.033748       0.123476    1.451179e+00  3.852845e+01   
2014-12-28    0.094198       2.645496    4.249335e+01  5.833801e-01   
2014-12-29    0.005964       3.041667    2.898325e+00  9.359547e-02   
2014-12-30    0.000028       1.131412    2.516381e-01  1.715028e-01   
2014-12-31    0.532747       2.370896    1.047835e-13  4.437736e+00   

model           BCC-ESM1  CMCC-CM2-HR4  CMCC-CM2-SR5  CMCC-ESM2    CanESM5  \
time                                                                         
1889-01-01  1.107466e-03  1.141054e+01  3.322009e-08   2.668800   1.321215   
1889-01-02  1.015323e-04  4.014984e+00  1.312700e+00   0.946211   2.788724   
1889-01-03  1.760345e-05  9.660565e+00  9.103720e+00   0.431999   0.003672   
1889-01-04  1.808932e-01  3.951528e+00  1.317160e+01   0.368693   0.013578   
1889-01-05  4.708019e-09  2.766362e+00  1.822940e+01   0.339267   0.002468   
...                  ...           ...           ...        ...        ...   
2014-12-27  2.061717e-03  8.179260e-09  1.171263e-02   0.090786  59.895053   
2014-12-28  5.939502e-09  8.146937e-01  4.938899e-01   0.000000   0.512632   
2014-12-29  2.000051e-08  2.532205e-01  1.306046e+00   0.000002  37.169669   
2014-12-30  7.191735e-05  8.169252e-02  1.722262e-01   0.788577   7.361246   
2014-12-31  2.863683e-01  6.343592e+00  6.368303e-01   0.442130   0.306608   

model       EC-Earth3-Veg-LR  ...  MPI-ESM-1-2-HAM  MPI-ESM1-2-HR  \
time                          ...                                   
1889-01-01          1.515293  ...     4.244226e-13   1.390174e-13   
1889-01-02          4.771375  ...     4.409552e+00   1.222283e-01   
1889-01-03          4.233980  ...     2.269300e-01   3.762301e-01   
1889-01-04         15.252495  ...     2.344586e-02   4.214019e-01   
1889-01-05         11.920356  ...     4.270161e-13   1.879692e-01   
...                      ...  ...              ...            ...   
2014-12-27          5.071783  ...     4.726998e-13   1.326889e-01   
2014-12-28          1.578188  ...     4.609420e-13   1.644482e+00   
2014-12-29          1.565885  ...     2.016156e+01   1.506439e+00   
2014-12-30          0.025749  ...     9.420543e+00   6.242895e+00   
2014-12-31          0.106525  ...     1.031899e+01   4.765813e+01   

model       MPI-ESM1-2-LR    MRI-ESM2-0         NESM3    NorESM2-LM  \
time                                                                  
1889-01-01   6.537884e-05  3.445495e-06  1.576096e+01  4.759651e-05   
1889-01-02   1.049131e-13  4.791993e-09  3.675510e-01  4.350863e-01   
1889-01-03   9.758706e-14  6.912302e-01  1.562869e-01  9.561101e+00   
1889-01-04   7.060915e-03  3.835721e-02  2.472226e-07  5.301038e-01   
1889-01-05   4.504985e+00  3.506923e-07  1.949792e-13  1.460928e-10   
...                   ...           ...           ...           ...   
2014-12-27   1.827857e+00  6.912632e-03  2.171327e-03  1.620489e+00   
2014-12-28   7.242920e-01  2.836752e-03  1.344768e+01  2.391159e+00   
2014-12-29   1.049481e-01  8.137182e+00  2.547820e+01  1.987695e-12   
2014-12-30   1.245115e-01  9.305263e-03  4.192948e+00  2.150346e+00   
2014-12-31   3.274323e-01  6.854985e-11  2.067847e+00  2.349716e+01   

model       NorESM2-MM  SAM0-UNICON    TaiESM1   observed  
time                                                       
1889-01-01    2.451075     0.221324   2.257933   0.006612  
1889-01-02    0.477231     3.757179   2.287381   0.090422  

In [70]:
df_sydney

model       ACCESS-CM2  ACCESS-ESM1-5  AWI-ESM-1-1-LR   BCC-CSM2-MR  \
time                                                                  
1889-01-01    0.040427       1.814552    3.557934e+01  4.268112e+00   
1889-01-02    0.073777       0.303965    4.596520e+00  1.190141e+00   
1889-01-03    0.232656       0.019976    5.927467e+00  1.003845e-09   
1889-01-04    0.911319      13.623777    8.029624e+00  8.225225e-02   
1889-01-05    0.698013       0.021048    2.132686e+00  2.496841e+00   
...                ...            ...             ...           ...   
2014-12-27    0.033748       0.123476    1.451179e+00  3.852845e+01   
2014-12-28    0.094198       2.645496    4.249335e+01  5.833801e-01   
2014-12-29    0.005964       3.041667    2.898325e+00  9.359547e-02   
2014-12-30    0.000028       1.131412    2.516381e-01  1.715028e-01   
2014-12-31    0.532747       2.370896    1.047835e-13  4.437736e+00   

model           BCC-ESM1  CMCC-CM2-HR4  CMCC-CM2-SR5  CMCC-ESM2    CanESM5  \
time                                                                         
1889-01-01  1.107466e-03  1.141054e+01  3.322009e-08   2.668800   1.321215   
1889-01-02  1.015323e-04  4.014984e+00  1.312700e+00   0.946211   2.788724   
1889-01-03  1.760345e-05  9.660565e+00  9.103720e+00   0.431999   0.003672   
1889-01-04  1.808932e-01  3.951528e+00  1.317160e+01   0.368693   0.013578   
1889-01-05  4.708019e-09  2.766362e+00  1.822940e+01   0.339267   0.002468   
...                  ...           ...           ...        ...        ...   
2014-12-27  2.061717e-03  8.179260e-09  1.171263e-02   0.090786  59.895053   
2014-12-28  5.939502e-09  8.146937e-01  4.938899e-01   0.000000   0.512632   
2014-12-29  2.000051e-08  2.532205e-01  1.306046e+00   0.000002  37.169669   
2014-12-30  7.191735e-05  8.169252e-02  1.722262e-01   0.788577   7.361246   
2014-12-31  2.863683e-01  6.343592e+00  6.368303e-01   0.442130   0.306608   

model       EC-Earth3-Veg-LR  ...  MPI-ESM-1-2-HAM  MPI-ESM1-2-HR  \
time                          ...                                   
1889-01-01          1.515293  ...     4.244226e-13   1.390174e-13   
1889-01-02          4.771375  ...     4.409552e+00   1.222283e-01   
1889-01-03          4.233980  ...     2.269300e-01   3.762301e-01   
1889-01-04         15.252495  ...     2.344586e-02   4.214019e-01   
1889-01-05         11.920356  ...     4.270161e-13   1.879692e-01   
...                      ...  ...              ...            ...   
2014-12-27          5.071783  ...     4.726998e-13   1.326889e-01   
2014-12-28          1.578188  ...     4.609420e-13   1.644482e+00   
2014-12-29          1.565885  ...     2.016156e+01   1.506439e+00   
2014-12-30          0.025749  ...     9.420543e+00   6.242895e+00   
2014-12-31          0.106525  ...     1.031899e+01   4.765813e+01   

model       MPI-ESM1-2-LR    MRI-ESM2-0         NESM3    NorESM2-LM  \
time                                                                  
1889-01-01   6.537884e-05  3.445495e-06  1.576096e+01  4.759651e-05   
1889-01-02   1.049131e-13  4.791993e-09  3.675510e-01  4.350863e-01   
1889-01-03   9.758706e-14  6.912302e-01  1.562869e-01  9.561101e+00   
1889-01-04   7.060915e-03  3.835721e-02  2.472226e-07  5.301038e-01   
1889-01-05   4.504985e+00  3.506923e-07  1.949792e-13  1.460928e-10   
...                   ...           ...           ...           ...   
2014-12-27   1.827857e+00  6.912632e-03  2.171327e-03  1.620489e+00   
2014-12-28   7.242920e-01  2.836752e-03  1.344768e+01  2.391159e+00   
2014-12-29   1.049481e-01  8.137182e+00  2.547820e+01  1.987695e-12   
2014-12-30   1.245115e-01  9.305263e-03  4.192948e+00  2.150346e+00   
2014-12-31   3.274323e-01  6.854985e-11  2.067847e+00  2.349716e+01   

model       NorESM2-MM  SAM0-UNICON    TaiESM1   observed  
time                                                       
1889-01-01    2.451075     0.221324   2.257933   0.006612  
1889-01-02    0.477231     3.757179   2.287381   0.090422  

Finally our dataset is ready for training model of rainfall prediction with machine learning.

In [71]:
df_sydney.to_csv("s3://mds-s3-student65/output/ml_data_SYD.csv", storage_options=aws_credentials)

## Optional


Want to check out how much time it will take to read a CSV file instead of parquet from s3?

For that, upload the CSV file (combined_model_data.csv.zip )to S3 and try to read it instead of parquet.

You probably will get the answer for WHY parquet? :)


In [75]:
files_to_dl = ["combined_model_data.csv.zip"]  
for file in files:
    if file["name"] in files_to_dl:
        os.makedirs(output_directory, exist_ok=True)
        urlretrieve(file["download_url"], output_directory + file["name"])

In [77]:
with zipfile.ZipFile(os.path.join(output_directory, "combined_model_data.csv.zip"), 'r') as f:
    f.extractall(output_directory)

#### Reading in the `combined_model_data.csv` from s3

In [8]:


%%time
%%memit
df_csv = pd.read_csv("s3://mds-s3-student65/combined_model_data.csv/combined_model_data.csv", storage_options=aws_credentials)


peak memory: 12556.19 MiB, increment: 9176.34 MiB
CPU times: user 2min 8s, sys: 13.2 s, total: 2min 21s
Wall time: 13min 59s


In [9]:
df_csv.head()

time    lat_min    lat_max   lon_min   lon_max  \
0  1889-01-01 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
1  1889-01-02 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
2  1889-01-03 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
3  1889-01-04 12:00:00 -35.439867 -33.574619  141.5625  143.4375   
4  1889-01-05 12:00:00 -35.439867 -33.574619  141.5625  143.4375   

   rain (mm/day)            model  
0   4.244226e-13  MPI-ESM-1-2-HAM  
1   4.217326e-13  MPI-ESM-1-2-HAM  
2   4.498125e-13  MPI-ESM-1-2-HAM  
3   4.251282e-13  MPI-ESM-1-2-HAM  
4   4.270161e-13  MPI-ESM-1-2-HAM

In [11]:
df_csv.shape

(62513863, 7)

Indeed it takes more time reading from a csv file compared to the partitioned parquet file. Here is a brief comparison between the two approaches in terms of time taken to read in the data.

- **Partioned Parquet file**:
    - CPU times: user 10.2 s, total: 16 s
    - Wall time: 30.5 s
- **Combined CSV file**
    - CPU times: user 2min 8s, total: 2min 21s
    - Wall time: 13min 59s